src/telit_he910.cpp

```diff
@@ -0,0 +1,2102 @@
+					}
+				
+					break;
+			}
+		
+			break;
+			
+		case CLOSE_PDP:
+		
+			// deactivate data session (just in case the network thinks we still have an active PDP context)
+			if(closePDPContext())
+			{
+				subState = 0;
+				state = OPEN_PDP_DELAY;
+			}
+			else
+			{	
+				subState = 0;
+				state = POWER_OFF;
+			}
+
+			break;
+			
+		case OPEN_PDP_DELAY:
+		
+			switch(subState)
+			{
+				case 0:
+				
+					timer = uptimeMs() + 1000;
+					subState = 1;
+					
+					break;
+					
+				case 1:
+
+					if(uptimeMs() > timer)
+					{
+						state = OPEN_PDP;
+						subState = 0;
+					}
+					
+					break;
+			}
+			
+			break;
+			
+		case OPEN_PDP:
+		
+			// activate data session
+			if(openPDPContext())
+			{
+				pdp_counter = 0;
+				state = READY;
+			}
+			else
+			{	
+				if(pdp_counter < PDP_MAX_ATTEMPTS)
+				{
+					pdp_counter++;
+					state = CLOSE_PDP;
+				}
+				else
+				{
+					pdp_counter = 0;
+					state = POWER_OFF;
+				}
+			}
+		
+			break;
+			
+		case READY:
+		
+			switch(subState)
+			{
+				case 0:
+				
+					if(getNetworkConnectionStatus(&connectStatus), (connectStatus != REGISTERED_HOME && connectStatus != REGISTERED_ROAMING))
+					{
+						debug("Modem has lost network connection");
+						connect = false;
+						state = WAIT_FOR_NETWORK;
+					}
+					else if(getPDPContext(&pdp_connected), !pdp_connected)
+					{
+						debug("Modem has lost data session");
+						connect = false;
+						state = CLOSE_PDP;
+					}
+					else
+					{
+						connect = true;
+						timer = uptimeMs() + 1000;
+						subState = 1;
+					}
+				
+					break;
+					
+				case 1:
+				
+					if(uptimeMs() > timer)
+						subState = 0;
+				
+					break;
+			}
+		
+			break;
+			
+		default:
+		
+			state = POWER_OFF;
+		
+			break;
+	}
+	
+	return state;
+}
+
+void openxc::telitHE910::deinitialize() {
+	state = POWER_OFF;
+	setPowerState(false);
+}
+
+bool openxc::telitHE910::connected(TelitDevice* device) {
+
+	return connect;
+
+}
+
+static bool autobaud(openxc::telitHE910::TelitDevice* device) {
+
+	bool rc = false;
+	unsigned int i = 0;
+
+	// loop through a set of baud rates looking for a valid response
+	for(i = 0; i < 3; ++i)
+	{
+		// set local baud rate to next attempt
+		uart::changeBaudRate(device->uart, bauds[i]);
+		
+		// attempt set the remote baud rate to desired (config.h) value
+		if(openxc::telitHE910::setBaud(UART_BAUD_RATE) == true)
+		{
+			// match local baud to desired value
+			uart::changeBaudRate(device->uart, UART_BAUD_RATE);
+			
+			// we're good!
+			rc = true;
+			break;
+		}
+	}
+	
+	return rc;
+
+}
+
+/*MODEM AT COMMANDS*/
+
+bool openxc::telitHE910::saveSettings() {
+
+	bool rc = true;
+	
+	if(sendCommand(telitDevice, "AT&W0\r\n", "\r\n\r\nOK\r\n", 1000) == false)
+	{
+		rc = false;
+		goto fcn_exit;
+	}
+	
+	fcn_exit:
+	return rc;
+
+}
+
+bool openxc::telitHE910::setBaud(unsigned int baudRate) {
+
+	bool rc = true;
+	char command[32] = {};
+	
+	sprintf(command, "AT+IPR=%u\r\n", baudRate);
```

Comment:
If this uses AT-style commands, I wonder if you could build support for it into https://github.com/openxc/at-commander since we already use that to control the RN-4x module for Bluetooth.